linux/NvidiaJetson.c

```diff
@@ -0,0 +1,321 @@
+appropriately). The GpuFilter shows only the processes which currently utilize GPU (i.e. highly extended
+nvmap/iovmm/clients table). It is achieved by the filtering machinery associated with ProcessTable::pidMatchList.
+The code below constructs GPU_PID_MATCH_LIST hash table, then actionToggleGpuFilter either stores it to the
+ProcessTable::pidMatchList or restores old value of ProcessTable::pidMatchList.
+
+The separate LinuxProcess's PROCESS_FLAG_LINUX_GPU_JETSON (or something ...) flag isn't added for GPU_MEM, because
+currently the functionality of population LinuxProcess::gpu_mem is shared with the GPU consumers filter construction.
+So, even if GPU_MEM field is not activated, the filter showing GPU consumers should work. This kind of architecture is
+chosen intentially since it saves memory for the hash table GPU_PID_MATCH_LIST (which is now actually a set), and therefore
+increases performance. All other approaches convert GPU_PID_MATCH_LIST to a true key/value storage (key = pid,
+value = gpu memory allocated) with further merge code.
+
+*/
+
+
+/* global paths per each sensor */
+char CPU_TEMP_SENSOR_PATH[64];
+char GPU_TEMP_SENSOR_PATH[64];
+char GPU_FREQ_SENSOR_PATH[64];
+char GPU_LOAD_SENSOR_PATH[64];
+
+
+/* #define DEBUG_ME 1 */
+#define MAX_GPU_PROCESS_COUNT 256UL
+static Hashtable *GPU_PID_MATCH_LIST = NULL;
+
+
+static void NVidiaJetsonHashtableFunctor_ResetGpuMem(ATTR_UNUSED ht_key_t key, void* val, ATTR_UNUSED void* userData) {
+   LinuxProcess *lp = val;
+   lp->gpu_mem = 0;
+}
+
+
+Hashtable *NvidiaJetson_GetPidMatchList(void) {
+   return GPU_PID_MATCH_LIST;
+}
+
+
+void NvidiaJetson_LoadGpuProcessTable(Hashtable *pidHash) {
+   static int IsRootUser = -1;
+
+   /* needs root privileges */
+   if (!IsRootUser) return;
+
+   /* first time function is called */
+   if (IsRootUser == -1) {
+#ifdef DEBUG_ME
+      IsRootUser = 1;
+#else
+      IsRootUser = geteuid() == 0;
+#endif
+      if (!IsRootUser) return;
+
+      GPU_PID_MATCH_LIST = Hashtable_new(MAX_GPU_PROCESS_COUNT, false);
+   }
+
+#ifdef DEBUG_ME
+   FILE *input = fopen("/home/philimonov/tmp/test.file", "r");
+#else
+   FILE *input = fopen("/sys/kernel/debug/nvmap/iovmm/clients", "r");
+#endif
+   if (input == NULL) return;
+
+   /* reset all knowledge about GPU allocations */
+   Hashtable_foreach(pidHash, NVidiaJetsonHashtableFunctor_ResetGpuMem, NULL);
+   Hashtable_clear(GPU_PID_MATCH_LIST);
+
+   /* construct new knowledge regarding GPU allocations */
+   static const char sentinel = -128;
+   static const size_t line_sz = 256;
+
+   char line[line_sz], *last = &line[sizeof(line)-1]; *last = sentinel;
+   char *saveptr, *endptr;
+   while(fgets(line, sizeof(line), input)) {
+      if(String_startsWith(line, "CLIENT")) continue; // skip header
+      if(String_startsWith(line, "total")) break; //final line, skip for now
+      if(*last == '\0') {
```

Comment:
Space after keywords …

```suggestion
   while (fgets(line, sizeof(line), input)) {
      if (String_startsWith(line, "CLIENT")) // skip header
         continue;
      if (String_startsWith(line, "total")) //final line, skip for now
         break;

      if (*last == '\0') {
```